doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,146 @@
+The setting ``profiling: True`` tells GHC
+to insert performance measuring code into your application in general,
+Further in-depth information on profiling with GHC and its compiler options can be found in the
+`GHC profiling guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html>`__
+
+.. note::
+
+    While `cabal.project <cabal-project-file>`__ is intended for long-time settings
+    that are useful to store in Git, `cabal.project.local` is for short-lived, local experiments
+    (like profiling) that, in general, shouldn't be committed to Git.
+
+Second, run your application with the right runtime system flags and let it create a profiling report:
+
+.. code-block:: console
+
+    $ cabal run my-app +RTS -pj -RTS
+    <app builds, runs and finishes>
+
+When the application finishes, a profiling JSON report (due to option ``-pj``)
+is written to a ``<app-name>.prof`` file, i.e. ``my-app.prof``, in the current directory.
+
+.. note::
+
+    Different report formats can be generated by using different RTS flags. Some useful ones are:
+
+    - ``-p`` for a GHC's own
+      `standard report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#cost-centres-and-cost-centre-stacks>`__
+      ``<app-name>.prof``, which can be visualized with `profiteur <https://github.com/jaspervdj/profiteur>`__
+      or `ghcprofview <https://github.com/portnov/ghcprofview-hs>`__.
+    - ``-pj`` for a
+      `JSON report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#json-profile-format>`__
+      ``<app-name>.prof``, which can be visualized with `Speedscope <https://speedscope.app>`__.
+    - ``-l-p`` for a binary
+      `"eventlog" report <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#rts-eventlog>`__
+      ``<app-name>.eventlog``, which contains a lot more details and can show you resource usage over time, and can
+      be converted to JSON with `hs-speedscope <https://github.com/mpickering/hs-speedscope>`__
+      to be visualized with `Speedscope <https://speedscope.app>`__.
+
+Finally, visualize this JSON report ``my-app.prof`` and analyze it for performance bottlenecks.
+One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is
+`Speedscope <https://speedscope.app>`__,
+which runs in the browser and can open this JSON file directly.
+See the
+`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
+on how to optimize your code based on the profiling results afterwards.
+
+So far, we've only used a single GHC compiler option to enable profiling
+in general for your application.
+However, you can pass a lot more compiler flags to GHC via the ``cabal.project.local`` file:
+
+.. code-block:: cabal
+
+    profiling: True
+    profiling-detail: late-toplevel
+    program-options
+      ghc-options:
+        -fprof-auto
+        -fno-prof-count-entries
+        -fprof-auto-calls
```

Comment:
There is a bit of confusion here. `-fprof-auto` is the same as `profiling-detail: all-functions`, and `-fprof-auto-calls` is another one of these flags so you are specifying a (kind of) profiling detail thrice. I assume maybe they all are applied and you get all non-inline definitions + all call sites + all top-level after optimizations?

Also if you mention program-options, I would link to the appropriate place: https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#cfg-section-program-options-program-options

Reply:
It would be nice if the code blocks could contain links. I don't know if that is possible in RST.

Reply:
I don't think it is: the main point of a code block is to disable markup which might have a different meaning in code.

Reply:
I added a sentence about `program-options` below with a reference link.

Regarding those specific `ghc-options` i added them, because @Kleidukos suggested them (i can imagine these are a good default to promote), but i have never used them before. I added the link to the GHC explanations a bit below, which i wouldn't want to duplicate here.

Reply:
I'm unsure, I think `-fprof-auto` and `-fprof-auto-calls` will get in the way for optimizations, which is precisely what we would like to avoid with `-fprof-late`. I don't think these should be advertised together.

Reply:
I stand corrected. So, in context of late profiling, `-fprof-auto` just extends it beyond top-level definitions? If so, it makes sense to include it, especially if the Cabal manual is correct that our `late` is equivalent to `late-toplevel`, so there's no easy way to demand more than top-level definitions. I couldn't find in the cabal manual whether late or non-late profiling is the cabal default (GHC's is late, I think, so maybe it's inherited from the GHC in use?). Anyway, sorry for the noise and I'm glad this gets enough pairs of eyes.

Reply:
So is this resolved/okay?

Reply:
I think I would write:

> So far, we’ve only used a single cabal option to enable profiling in general for your application. Leaving [`profiling-detail`](https://cabal--10539.org.readthedocs.build/en/10539/cabal-project-description-file.html#cfg-field-profiling-detail) unspecified defaults to `profiling-detail: exported-functions` for libraries and `profiling-detail: toplevel-functions` for executables. 
>
> However, you can specify the profiling detail and provide more compiler flags to GHC (such as `-fno-prof-count-entries`) via the `cabal.project.local` file:
> ```
> profiling: True
> profiling-detail: late-toplevel
> program-options
>   ghc-options:
>     <further options>
> ```

Reply:
Sounds good, but I'd double check that we indeed did not follow GHC and did not make late profiling the default. Or maybe we insert the same GHC flags but now they mean late profiling?

Reply:
I applied this paragraph. If it's not accurate, please provide an alternative suggestions .

Reply:
I don't have a reasonable suggestion without code-diving, but fortunately others have commented, so let's focus there.  Thank you!